File: 2015/day20.c
#include <stdio.h>
#include <math.h>

int p1(int n) {
    int i, sum = 0, end = sqrt(n);
    for (i = 1; i <= end; i++) {
        if ((n % i) == 0) sum += ((n/i) != i ? i + (n/i) : i);
    }
    return sum * 10;
}

int p2(int n) {
    int i, sum = 0, end = sqrt(n);
    for (i = 1; i <= end; i++) {
        if ((n % i) == 0) sum += ((n/i) <= 50 ? i : 0) +
                                 (i     <= 50 && (n/i) != i ? (n/i) : 0);
    }
    return sum * 11;
}

int search1(int n) {
   int i, lo = n / 45, hi = n / 10;
   for (i = lo; i <= hi; i+=5) if (p1(i) >= n) { hi = i; break; }
   for (i = lo; i <= hi; i++)  if (p1(i) >= n) return i;
   return -1;
}

int search2(int n) {
   int i, lo = n / 43, hi = n / 11;
   for (i = lo; i <= hi; i+=5) if (p2(i) >= n) { hi = i; break; }
   for (i = lo; i <= hi; i++)  if (p2(i) >= n) return i;
   return -1;
}

int main() {
    int presents_target = 33100000;
    printf("2015 day 20 part 1: %d\n", search1(presents_target));
    printf("2015 day 20 part 2: %d\n", search2(presents_target));
}

File: 2015/day20.py
import math

def divisors(n):
    return (i for i in range(1, int(math.sqrt(n)) + 1) if (n % i) == 0)

# presents for house n = 10 * sum(positive factors of n)
def p1(n):
    return 10 * sum(((i if i == n//i else i + n//i) for i in divisors(n)))

# presents for house n = 11 * sum(positive factors of n where n/factor <= 50)
def p2(n):
    return 11 * sum(( (i    if n//i <= 50 else 0) +
                      (n//i if i    <= 50 and n//i != i else 0) for i in divisors(n)))

def search(f, n, l, h, step=1):
    for i in range(l, h+1, step):
        if f(i) >= n: return l, i
    return l, h

# to narrow the search space:
# - the slope of p1(x)/x is between 10 and ~45
# - the slope of p2(x)/x is between 11 and ~43
# - checking every 5th value eliminates ~250,000 candidates with ~8,000 tests

def search1(n):
    lo, hi = search(p1, n, n // 45, n // 10, 5)
    return search(p1, n, lo, hi)[1]

def search2(n):
    lo, hi = search(p2, n, n // 43, n // 11, 5)
    return search(p2, n, lo, hi)[1]

presents_target = 33100000
print("2015 day 20 part 1: %d" % search1(presents_target))
print("2015 day 20 part 2: %d" % search2(presents_target))
